docs/design/secure-config.md

```diff
@@ -0,0 +1,337 @@
+    abc123/
+      metadata.binpb
+      config.toml
+  workspaces/
+    def456/
+      config.toml
+      metadata.binpb
+my-repo/.jj/
+  workspace-id (contains "def456")
+  workspace-config.toml (unused by jj, details below)
+  repo/
+    repo-id (contains "abc123")
+    config.toml (unused by jj, details below)
+```
+
+`metadata.binpb` will refer to the following protobuf:
+
+```proto
+message Metadata {
+  // This is used to distinguish between copies and moves.
+  string path = 1;
+}
+```
+
+The function to load repository configuration, will roughly speaking, look like:
+```rust
+enum ConfigLoadError { NoRepoId, NoConfig, PathMismatch, }
+
+fn load_repo_config_path(repo: &Path) -> Result<PathBuf, ConfigLoadError> {
+  let repo_id = std::fs::read_to_string(repo.join("repo-id"))
+    .map_err(|_|Err(NoRepoId))?;
+  let repo_config_dir = config_dir.join("repos").join(repo_id);
+  let metadata = Metadata::decode(
+    std::fs::read(repo_config_dir.join("metadata.binpb")) .map_err(|_|
+    Err(NoConfig))? )?;
+  if metadata.path != repo {
+    return Err(PathMismatch)
+  }
+  Ok(repo_config_dir.join("config.toml"))
+}
+```
+
+#### Happy path
+
+Normally we will simply:
+1. Load the `repo-id` file
+2. Check that `$HOME/.config/jj/repos/$REPO_ID` exists
+3. Validate that the path in `$HOME/.config/jj/repos/$REPO_ID/metadata.binpb`
+   matches the repo's path
+4. Find the corresponding config in
+   `$HOME/.config/jj/repos/$REPO_ID/config.toml`
+5. Load that config.
+
+However, these steps can fail. This is how we will handle the errors.
+
+#### No repo-ID file
+
+##### Without config files
+
+If `$repo/.jj/repo/config.toml` does not exist, then we don't need to generate a
+repo-ID (since it's only relevant for finding the correct place to read
+configs). We can just return an empty repo config.
+
+Note that the expected way to generate a repo-ID would be for the user to either
+run `jj config edit` or `jj config path`.
+
+##### With config files
+
+If the config file exists, then this corresponds to a legacy repo. To preserve
+backwards compatibility, we will introduce a period of auto-migration. The
+current plan is 12 jj versions (approximately 1 year). During this period, if a
+repo ID has not yet been generated, we will silently perform the following
+(order matters, to ensure failure halfway through doesn’t affect things):
+
+1.  Generate a repo ID `abc123`
+2.  Create `$HOME/.config/jj/repo/abc123/metadata.binpb`
+3.  Create `$HOME/.config/jj/repo/abc123/config.toml` as a copy of the original
+    config file
+4.  Atomically generate a `repo-id` file containing `abc123`
+5.  Remove the original config file
+6.  For the user's convenience, and for older version of jj, we:
+    *   Try to symlink the old config to the new config
+    *   If this fails (symlinks don't play nice on windows), we replace it with
+        the same file content, with an extra comment at the top telling the user
+        not to edit the file, and set it to readonly.
```

Comment:
Since this may help attacker know both `repo-id` and the associated config content, we might have to delete the copied config file after the migration period is over.

Also note that `zip` by default follows symlinks. `tar` wouldn't, though. (I have no idea about Windows tooling.)

Reply:
> Since this may help attacker know both `repo-id` and the associated config content, we might have to delete the copied config file after the migration period is over.

An attacker who you send a repo to will already know the repo-id, since the repo-id file exists, so I'm not worried about that part. Knowing the config content is an issue though. I don't particularly like deleting it, but I do agree it may be our best option.